platform/core-api/src/org/jetbrains/concurrency/ParallelUtil.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2000-2015 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jetbrains.concurrency;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.util.registry.Registry;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.concurrent.*;
+
+/**
+ * Utilities for executing small jobs in parallel.
+ */
+public class ParallelUtil {
+  private static final Logger LOG = Logger.getInstance(ParallelUtil.class);
+  private static final int ourReasonableThreadPoolSize = Registry.intValue("core.pooled.threads");
+  private static final ThreadPoolExecutor ourExecutorService =
```

Comment:
Please check BoundedTaskExecutor, it can work above pooled threads of Application